daffodil-core/src/main/scala/org/apache/daffodil/dsom/ModelGroup.scala

```diff
@@ -57,41 +58,54 @@ object ModelGroupFactory {
       lexicalParent.schemaDefinitionError("Model group circular definitions. Group references, or hidden group references form a loop.")
     } else {
       val moreNodesAlreadyTrying = nodesAlreadyTrying + child
+      val res = // memoizedApply(child)(lexicalParent)(position)(isHidden)(nodesAlreadyTrying)
+        nonMemoizedApply(child, lexicalParent, position, isHidden, nodesAlreadyTrying)
+      res
+    }
+  }
 
-      val childModelGroup: ModelGroup = child match {
-        case <sequence>{ _* }</sequence> => {
-          val seq = new Sequence(child, lexicalParent, position)
-          if (seq.hiddenGroupRefOption.isDefined) {
-            //
-            // construct the group ref XML, then recursively process that,
-            // but set flag so it will be hidden.
-            //
-            val hgrXML = seq.hiddenGroupRefXML
-            ModelGroupFactory(hgrXML, lexicalParent, position, true, moreNodesAlreadyTrying)
-          } else {
-            seq
-          }
-        }
-        case <choice>{ _* }</choice> => new Choice(child, lexicalParent, position)
-        case <group>{ _* }</group> => {
-          val pos = lexicalParent match {
-            case ct: ComplexTypeBase => 1
-            case mg: ModelGroup => position
-            case gd: GlobalGroupDef => position
-          }
-          val isH = isHidden || lexicalParent.isHidden
-          val groupRefFactory = new GroupRefFactory(child, lexicalParent, pos, isH)
-          val groupRefInstance = groupRefFactory.groupRef
-          groupRefInstance.asModelGroup
+  private lazy val memoizedApply =
+    Memoize1((child: Node) =>
+      Memoize1((lexicalParent: SchemaComponent) =>
+        Memoize1((position: JInt) =>
+          Memoize1((isHidden: JBoolean) =>
+            (nodesAlreadyTrying: Set[Node]) =>
+              nonMemoizedApply(child, lexicalParent, position, isHidden, nodesAlreadyTrying)))))
+
+  private def nonMemoizedApply(child: Node, lexicalParent: SchemaComponent, position: JInt, isHidden: Boolean,
```

Comment:
Can we improve the name of memoizedApply and nonMemoizedApply (and probably add a comment).

I get the sense that what we are "applying" is lexical component to a particular context. I assume the intent here is that it will allow us to still specialize terms when their lexical context requires it, but allows for us to use memoization to avoid re-computing terms that do not require it.

One issue I see with this approach (which may be unavoidable) is that it is highly attractive for developers to add new parameters without giving much thought to exponential blowup. On that note, I see that we are dependent on lexicalParent, which seems like an overly broad dependency .

Reply:
I removed all the memoize stuff, as it wasn't being used, and your observations are very apropos. Memoize approaches work best when the equality comparison work involved is well understood to be either pointer-based equality or operating on simple values or simple tuples of values. 

This was half baked, and not in use as yet, so I removed it for now.